### PR TITLE
Feature/signing in

### PR DESCRIPTION
**Pull request by patrykmierunski**

### CODE REVIEW

File: src/main/java/pl/mantiscrab/budgetr/config/H2Config.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import org.h2.tools.Server;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Profile;
+
+import java.sql.SQLException;
+
+@Configuration
+@Profile({"dev", "test"})
+class H2Config {
```
Comment (greencashew) [inline src/main/java/pl/mantiscrab/budgetr/config/H2Config.java:12]: It's fully fine to have this canfig in here bu I thought that autoconfiguration is doing exactly the same

File: src/main/java/pl/mantiscrab/budgetr/config/SecurityFilterChainConfig.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                .requestMatchers(PathRequest.toH2Console()).permitAll()
+                .anyRequest().authenticated());
+        http.httpBasic();
+        http.csrf().disable();
+        http.headers().frameOptions().sameOrigin();
+        return http.build();
+    }
+
+    @Bean
+    @Profile("test")
```
Comment (greencashew) [inline src/main/java/pl/mantiscrab/budgetr/config/SecurityFilterChainConfig.java]: `test` profile should never appear in src code.

You can write your own configuration under the test packages. But like your idea you wanted to have everything in single place.

File: src/main/java/pl/mantiscrab/budgetr/registration/InMemoryUserRepository.java [hunk 3]
```diff
@@ -1,22 +1,21 @@
 package pl.mantiscrab.budgetr.registration;
 
+import org.springframework.dao.DataIntegrityViolationException;
+import pl.mantiscrab.budgetr.InMemoryDummyCrudRepository;
+
 import java.util.Objects;
 import java.util.Optional;
 
-class InMemoryUserRepository extends InMemoryDummyCrudRepository<User, Long> implements UserRepository{
+class InMemoryUserRepository extends InMemoryDummyCrudRepository<User, String> implements UserRepository{
```
Comment (greencashew) [inline src/main/java/pl/mantiscrab/budgetr/registration/InMemoryUserRepository.java]: What is the purpose of having both h2 and this repositories?

File: src/main/java/pl/mantiscrab/budgetr/registration/User.java [hunk 4]
```diff
@@ -15,17 +16,27 @@
...
     @Column(nullable = false)
     @NotBlank
     private String password;
 
+    @Column(nullable = false)
+    private boolean enabled;
+
+    @OneToMany(mappedBy = "user", cascade = CascadeType.ALL)
+    @ToString.Exclude
+    private Set<Authority> authorities;
```
Comment (greencashew) [inline src/main/java/pl/mantiscrab/budgetr/registration/User.java:38]: don't really get the idea about the authorities

File: src/main/java/pl/mantiscrab/budgetr/registration/UserController.java [hunk 5]
```diff
@@ -25,6 +24,14 @@ ResponseEntity<UserDto> register(@Valid @RequestBody UserRegisterDto userRegiste
         return ResponseEntity.ok().body(userDto);
     }
 
+    //    TODO: replace this endpoint in tests with one related
+    //    TODO: to business logic and remove
+    @GetMapping("/username")
```
Comment (greencashew) [inline src/main/java/pl/mantiscrab/budgetr/registration/UserController.java]: User controller not restful :dagger: 

POST /user
PUT /user
PATCH /user
DELETE /user
GET /user

File: src/main/java/pl/mantiscrab/budgetr/registration/UserService.java [hunk 6]
```diff
@@ -25,15 +26,17 @@ private void throwExceptionIfUserAlreadyExists(UserRegisterDto dto) {
...
-    private User mapUserDtoToUser(UserRegisterDto dto) {
-        UserRegisterDto dtoWithEncodedPassword = encodePassword(dto);
-        User user = UserMapper.userFromUserRegisterDto(dtoWithEncodedPassword);
+    private User createUser(UserRegisterDto dto) {
+        String encodedPassword = passwordEncoder.encode(dto.password());
+        User user = User.builder()
+                .email(dto.email())
+                .username(dto.username())
+                .password(encodedPassword)
+                .authorities(new HashSet<>())
```
Comment (greencashew) [inline src/main/java/pl/mantiscrab/budgetr/registration/UserService.java]: better put nothing in here?

File: src/main/resources/application-test.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+spring:
```
Comment (greencashew) [inline src/main/resources/application-test.yml]: no test profile w source

**Review by greencashew — CHANGES_REQUESTED** [review]

File: src/main/java/pl/mantiscrab/budgetr/config/H2Config.java (hunk 1 above)
Comment (patrykmierunski) [reply src/main/java/pl/mantiscrab/budgetr/config/H2Config.java:12]: It doesn't. For further explanations: https://www.youtube.com/watch?v=HlvF2TlA1Yo
Tl;dr : without it IntelliJ creates it's own instance of H2 database while connecting, so you don't have access to "application's" database

File: src/main/java/pl/mantiscrab/budgetr/registration/User.java (hunk 4 above)
Comment (patrykmierunski) [reply src/main/java/pl/mantiscrab/budgetr/registration/User.java:38]: Application must know if user sending request is ordinary user or admin, moderator, superadmin etc. Right now I only have ordinary users, but I might have some different roles in the future. Besides, as far as I understand it, Spring Security requires users to have "some" role/authority.

File: src/main/java/pl/mantiscrab/budgetr/registration/InMemoryUserRepository.java (hunk 3 above)
Comment (patrykmierunski) [reply src/main/java/pl/mantiscrab/budgetr/registration/InMemoryUserRepository.java]: I don't understand the question. In memory repositories are for unit tests, and H2 for integration

File: src/main/java/pl/mantiscrab/budgetr/registration/UserService.java (hunk 6 above)
Comment (patrykmierunski) [reply src/main/java/pl/mantiscrab/budgetr/registration/UserService.java]: Did you mean ".authorities(new HashSet<>())" ? I had to, because otherwise it might result in "NullPointerException"

File: src/main/java/pl/mantiscrab/budgetr/registration/User.java (hunk 4 above)
Comment (greencashew) [reply src/main/java/pl/mantiscrab/budgetr/registration/User.java:38]: :+1: I meant about way of implementing authorities :-) But I think it's quite good option so far :-)

File: src/main/java/pl/mantiscrab/budgetr/registration/UserController.java [hunk 8]
```diff
@@ -25,6 +24,14 @@ ResponseEntity<UserDto> register(@Valid @RequestBody UserRegisterDto userRegiste
         return ResponseEntity.ok().body(userDto);
     }
 
+    //    TODO: replace this endpoint in tests with one related
+    //    TODO: to business logic and remove
+    @GetMapping("/user")
+    ResponseEntity<String> getUsername() {
```
Comment (greencashew) [inline src/main/java/pl/mantiscrab/budgetr/registration/UserController.java:30]: As i said, read about restful and how controller should look properly

File: src/test/java/pl/mantiscrab/budgetr/registration/SecurityFilterChainTestConfig.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.web.SecurityFilterChain;
+import pl.mantiscrab.budgetr.config.SecurityFilterChainConfig;
+
+@Configuration
+class SecurityFilterChainTestConfig {
+
+    @Bean
+    @Profile("test")
+    public SecurityFilterChain filterChainTest(HttpSecurity http) throws Exception {
```
Comment (greencashew) [inline src/test/java/pl/mantiscrab/budgetr/registration/SecurityFilterChainTestConfig.java:15]: :+1:

**Review by greencashew — APPROVED** [review]

**Comment by greencashew** [comment]
In overall, very good work :-) !
